### PR TITLE
gh-127599: Fix _Py_RefcntAdd missing calls to _Py_INCREF_STAT_INC/_Py_INCREF_IMMORTAL_STAT_INC

### LINKED ISSUE #127599: `_Py_RefcntAdd` doesn't increment ref count stats
# Bug report

### Bug description:

Hi,

I am unsure if this intentional or not so apologies if this turns out not to be a bug. 

The code for `_Py_RefcntAdd` does not include `_Py_INCREF_STAT_INC`. This appears to mean the statistics do not track these additions to the reference count, which are increments (although not increments by `1`, they are increasing the ref count in a single operation).

https://github.com/python/cpython/blob/ad9d059eb10ef132edd73075fa6d8d96d95b8701/Include/internal/pycore_object.h#L121

I believe this may be an oversight because this stat should be tracked, and also because the corresponding decrement ref count code does call `_Py_DECREF_STAT_INC` (e.g. https://github.com/python/cpython/blob/ad9d059eb10ef132edd73075fa6d8d96d95b8701/Include/internal/pycore_object.h#L383 )

I have checked the callers of `_Py_RefcntAdd` (in `listobject.c` and `tupleobject.c`) and they do not separately perform the stats tracking.

Is this an issue or is there a reason this ref count increment is excluded?

Thanks,
Ed

### CPython versions tested on:

3.12, 3.13

### Operating systems tested on:

Linux, macOS

### Linked PRs
* gh-127717
* gh-127963
* gh-128712
* gh-128713

### PR DESCRIPTION
**Pull request by EdNutting**
Fixes #127599 - `_Py_RefcntAdd` doesn't increment ref count stats

* Issue: gh-127599

### CODE REVIEW

File: Misc/NEWS.d/next/Core_and_Builtins/2024-12-07-13-06-09.gh-issue-127599.tXCZb_.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fixes ``_Py_RefcntAdd`` missing calls to ``_Py_INCREF_STAT_INC`` / ``_Py_INCREF_IMMORTAL_STAT_INC``
```
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-12-07-13-06-09.gh-issue-127599.tXCZb_.rst]: I'd suggest having a changelog a bit more general, e.g., "Fix statistics for object's reference count increment (possibly by more than one)" or something like this (I don't have a good formulation for this so we can bikeshed a bit more). Unless we document those macros and `_Py_RefcntAdd`, I don't think it helps the user reading the changelog entry (AFAIK, `_Py_RefcntAdd` and the macros are only internal).

Comment (EdNutting) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-12-07-13-06-09.gh-issue-127599.tXCZb_.rst]: I've pushed an alternative description that captures the more abstract information. Let me know if you think it's better.

File: Misc/NEWS.d/next/Core_and_Builtins/2024-12-07-13-06-09.gh-issue-127599.tXCZb_.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Fix statistics for increments of object reference counts (in particular, when
+a reference count was increased by more than 1 in a single operation).
```
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-12-07-13-06-09.gh-issue-127599.tXCZb_.rst:2]: We can keep this one and drop the sentence about _Py_RefcntAdd.

Comment (EdNutting) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-12-07-13-06-09.gh-issue-127599.tXCZb_.rst:2]: Thanks, done.

**Review by Eclips4 — APPROVED** [review]
This looks correct, thank you!
If @picnixz agrees with wording (comment on `pystats.h` and NEWS entry), I'll merge it.

**Review by picnixz — APPROVED** [review]

**Comment by EdNutting** [comment]
@Eclips4 Looks like this is ready to merge - thank you both!

**Comment by Eclips4** [comment]
3.12 does not have a `_Py_INCREF_IMMORTAL_STAT_INC`. I do not think it is worth the effort to make it work on 3.12, so let's just backport it to a 3.13 branch.

**Comment by picnixz** [comment]
If the backports are not hard to do, let's backport it. I see it as a bug fix personally. But if the backports are really hard to do and prone to introducing bugs, then let's not do it and treat it as a feature.

**Comment by EdNutting** [comment]
My view is this is a bug fix not a feature. I don't hold a strong view on whether that means it should be backported or not.

Given that metrics are the core of performance analysis, anyone attempting to analyse 3.12/3.13 in their deployments might end up slightly off the mark without this fix.

My team and I are working with 3.12 internally, so I'll end up developing an internal patch over the New Year anyway. I'll see how much work it is to turn that into a proper backport patch for 3.12/3.13 and let you know.

**Comment by Eclips4** [comment]
Thank you both for sharing your thoughts.
Though, the problem with 3.12 and 3.13 is that they don't have `_Py_INCREF_IMMORTAL_STAT_INC`, so we need to use only an `_Py_INCREF_STAT_INC`. If you are with that - please submit a backport PR's to 3.12 && 3.13. I would be happy to merge it.

**Pull request merged**